include/exadg/operators/solution_projection_between_triangulations.h

```diff
@@ -0,0 +1,134 @@
+};
+
+/**
+ * Utility function to collect integration points in the particular sequence they are encountered
+ * in.
+ */
+template<int dim, int n_components, typename Number>
+std::vector<dealii::Point<dim>>
+collect_integration_points(
+  dealii::MatrixFree<dim, Number, dealii::VectorizedArray<Number>> const & matrix_free,
+  unsigned int const                                                       dof_index,
+  unsigned int const                                                       quad_index);
+
+/**
+ * Utility function to compute the right hand side of a projection (mass matrix solve)
+ * with values given in integration points in the particular sequence they are encountered in.
+ */
+template<int dim, int n_components, typename Number, typename VectorType>
+VectorType
+assemble_projection_rhs(
+  dealii::MatrixFree<dim, Number, dealii::VectorizedArray<Number>> const & matrix_free,
+  std::vector<
+    typename dealii::FEPointEvaluation<n_components, dim, dim, Number>::value_type> const &
+                     values_source_in_q_points_target,
+  unsigned int const dof_index,
+  unsigned int const quad_index);
+
+/**
+ * Utilitiy function to project vectors from a source to a target triangulation via
+ * matrix-free mass operator evaluation and preconditioned CG solver.
+ */
+template<int dim, typename Number, int n_components, typename VectorType>
+void
+project_vectors(
+  std::vector<VectorType *> const &                                        source_vectors,
+  dealii::DoFHandler<dim> const &                                          source_dof_handler,
+  std::shared_ptr<dealii::Mapping<dim> const> const &                      source_mapping,
+  std::vector<VectorType *> const &                                        target_vectors,
+  dealii::MatrixFree<dim, Number, dealii::VectorizedArray<Number>> const & target_matrix_free,
+  dealii::AffineConstraints<Number> const &                                constraints,
+  unsigned int const                                                       dof_index,
+  unsigned int const                                                       quad_index,
+  GridToGridProjectionData<dim> const &                                    data);
+
+/**
+ * Utility function to perform matrix-free grid-to-grid projection. We assume we only have a single
+ * `dealii::FiniteElement` per `dealii::DoFHandler`.
+ */
+template<int dim, typename VectorType>
+void
+grid_to_grid_projection(
+  std::vector<std::vector<VectorType *>> const &       source_vectors_per_dof_handler,
+  std::vector<dealii::DoFHandler<dim> const *> const & source_dof_handlers,
+  std::shared_ptr<dealii::Mapping<dim> const> const &  source_mapping,
+  std::vector<std::vector<VectorType *>> &             target_vectors_per_dof_handler,
+  std::vector<dealii::DoFHandler<dim> const *> const & target_dof_handlers,
+  std::shared_ptr<dealii::Mapping<dim> const> const &  target_mapping,
+  GridToGridProjectionData<dim> const &                data);
```

Comment:
Besides these general functions, shouldn't we also try to provide a function that can use an existing `MatrixFree` object for the mass matrix operator? The reason is that build a `MatrixFree::reinit()` object is expensive compared to a mass matrix solve in the usual case, for DG elements with cell-wise inverse mass matrices the cost difference is even larger.

Reply:
yes, good point! :+1: 
I will add another function that wraps the existing `grid_to_grid_projection()` and passes a `MatrixFree` object.